src/TypeScript/NodeJSLocator/NodeLocatorAggregator.cs

```diff
@@ -0,0 +1,98 @@
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
+ * Lesser General Public License for more details.
+ *
+ * You should have received a copy of the GNU Lesser General Public License
+ * along with this program; if not, write to the Free Software Foundation,
+ * Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301, USA.
+ */
+
+using System;
+using System.Collections.Generic;
+using System.ComponentModel.Composition;
+using System.Diagnostics;
+using SonarLint.VisualStudio.Integration;
+
+namespace SonarLint.VisualStudio.TypeScript.NodeJSLocator
+{
+    [Export(typeof(INodeLocator))]
+    [PartCreationPolicy(CreationPolicy.Shared)]
+    internal class NodeLocatorAggregator : INodeLocator
+    {
+        private readonly ILogger logger;
+        private readonly Func<string, Version> getNodeExeVersion;
+        private readonly IEnumerable<INodeLocator> nodeLocators;
+
+        [ImportingConstructor]
+        public NodeLocatorAggregator(INodeLocatorsProvider nodeLocatorsProvider, ILogger logger)
+            : this(nodeLocatorsProvider, logger, GetNodeVersion)
+        {
+        }
+
+        internal NodeLocatorAggregator(INodeLocatorsProvider nodeLocatorsProvider, ILogger logger, Func<string, Version> getNodeExeVersion)
+        {
+            this.logger = logger;
+            this.getNodeExeVersion = getNodeExeVersion;
+            nodeLocators = nodeLocatorsProvider.Get();
+        }
+
+        /// <summary>
+        /// Returns the absolute file path of a compatible `node.exe`, or null if no compatible version was found.
+        /// </summary>
+        public string Locate()
+        {
+            foreach (var nodeLocator in nodeLocators)
+            {
+                var nodeExePath = nodeLocator.Locate();
+
+                if (!string.IsNullOrEmpty(nodeExePath) && IsCompatibleVersion(nodeExePath))
+                {
+                    logger.WriteLine(Resources.INFO_FoundCompatibleVersion, nodeExePath);
+                    return nodeExePath;
+                }
+            }
+
+            logger.WriteLine(Resources.ERR_NoCompatibleVersion);
+            return null;
+        }
+
+        private bool IsCompatibleVersion(string nodeExePath)
+        {
+            var nodeVersion = getNodeExeVersion(nodeExePath);
+
+            logger.WriteLine(Resources.INFO_NodeVersion, nodeVersion.ToString(), nodeExePath);
+
+            if (nodeVersion.Major < 10)
+            {
+                logger.WriteLine(Resources.ERR_IncompatibleVersion);
+                return false;
+            }
+
+            if (nodeVersion.Major == 11)
+            {
+                logger.WriteLine(Resources.ERR_IncompatibleVersion11);
+                return false;
+            }
+
+            return true;
+        }
+
+        private static Version GetNodeVersion(string path)
+        {
+            var version = FileVersionInfo.GetVersionInfo(path);
```

Comment:
This is how it's done in the nodejstools:
https://github.com/microsoft/nodejstools/blob/275e85d5cd95cad9122f59a76f9e49bead66101b/Nodejs/Product/Nodejs/Nodejs.cs#L22

I can't mock `FileVersionInfo` so I had to mock the entire method. Unfortunately, system.io.abstractions doesn't provide a wrapper for it.

Reply:
I think it's worth putting a link to the Nodejs code in a comment for future reference (perhaps at the top of this class).

Reply:
Is this covered by any of tests? Could be a bit tricky since it directly accesses the file system; we wouldn't really want to have to create a new assembly with a known version on the fly, but if we use an existing assembly the test code would have to calculate the expected version in exactly the same way as the product code.

Reply:
Good suggestion; I ended up using the `System.ComponentModel.Composition.dll`. I couldn't use our own assemblies because it turns out that we don't populate the Product version fields, so it was always 0.0.0.